src/actions/write/foundationAction.ts

```diff
@@ -0,0 +1,276 @@
+              value[i.toString()] <= 0
+            ) {
+              throw new ContractError(
+                `Invalid value for fee ${i}. Must be an integer greater than 0`,
+              );
+            }
+          }
+        } else {
+          throw new ContractError(
+            `Invalid amount of fees.  Must be less than ${MAX_NAME_LENGTH}`,
+          );
+        }
+        break;
+      case 'createNewTier':
+        if (!Number.isInteger((value as ServiceTier).fee)) {
+          throw new ContractError('Fee must be a valid number.');
+        }
+        if (!Number.isInteger((value as ServiceTier).settings.maxUndernames)) {
+          throw new ContractError('Max undernames must be a valid number.');
+        }
+
+        (value as ServiceTier).id = SmartWeave.transaction.id;
+        break;
+      case 'setActiveTier':
+        // check that the tier number is valid
+        if (
+          !Number.isInteger((value as ActiveTier).tierNumber) ||
+          !ALLOWED_ACTIVE_TIERS.includes((value as ActiveTier).tierNumber)
+        ) {
+          throw new ContractError(DEFAULT_INVALID_TIER_MESSAGE);
+        }
+        // the tier must exist in the history before it can be set as an active tier
+        if (
+          !state.tiers.history.find(
+            (tier) => tier.id === (value as ActiveTier).tierId,
+          )
+        ) {
+          throw new ContractError(DEFAULT_INVALID_ID_TIER_MESSAGE);
+        }
+        break;
+      case 'delayedEvolve':
+        if (
+          typeof (value as DelayedEvolveInput).contractSrcTxId !== 'string' ||
+          !isValidArweaveBase64URL(
+            (value as DelayedEvolveInput).contractSrcTxId,
+          ) || // must be a valid arweave transaction ID
+          (value as DelayedEvolveInput).contractSrcTxId === state.evolve // must be new source code
+        ) {
+          throw new ContractError('Invalid contract evolution source code.');
+        }
+        if ((value as DelayedEvolveInput).evolveHeight) {
+          if (
+            !Number.isInteger((value as DelayedEvolveInput).evolveHeight) ||
+            (value as DelayedEvolveInput).evolveHeight -
+              +SmartWeave.block.height >=
+              MAX_ALLOWED_EVOLUTION_DELAY ||
+            (value as DelayedEvolveInput).evolveHeight -
+              +SmartWeave.block.height <
+              MINIMUM_ALLOWED_EVOLUTION_DELAY
+          ) {
+            throw new ContractError(
+              `Invalid contract evolution block height of ${
+                (value as DelayedEvolveInput).evolveHeight
+              }. Current height of ${+SmartWeave.block.height}`,
+            );
+          } else {
+            (value as DelayedEvolveInput).evolveHeight =
+              +SmartWeave.block.height + MINIMUM_ALLOWED_EVOLUTION_DELAY;
+          }
+        }
+        break;
+      default:
+        throw new ContractError('Invalid action parameters.');
+    }
+
+    id = foundation.actions.length;
```

Comment:
hmm - i'm thinking we should use the interaction id `Smartweave.transaction.id` - the length of the array might be out of date if two interactions are in the same block height. lookups would likely be safer to fat fingering the index number